File: kernel/src/libc/kfree.c
#include "../../include/memory.h"
#include "../../include/stdlib.h"

void kfree(void* ptr) {
    HEAP_block* it = &heap_blocks[0];
    while (it) {
        if (it->vaddr == ptr) {
            memset(ptr, 0, it->size);
            it->free = true;
            return;
        }
        it = it->next;
    }
    if (it == NULL) return;
}

File: kernel/src/libc/kfree.zig

const HEAP_block = extern struct {
    vaddr: u32,
    size: u32,
    free: bool,
    next: ?*HEAP_block,
};
extern var heap_blocks: [*]HEAP_block;
extern fn memset(ptr: *anyopaque, value: i32, num: u32) callconv(.c) *anyopaque;

export fn kfree(ptr: ?*anyopaque ) void {
    if (ptr == null) return;

    var it: ?*HEAP_block = &heap_blocks[0];
    while (it) |block| {
        if (block.vaddr == @intFromPtr(ptr.?)) {
            _ = memset(ptr.?, 0, block.size);
            block.free = true;
            return;
        }
        it = block.next;
    }
}
